src/lib/form/validators.ts

```diff
@@ -0,0 +1,153 @@
+import { AbstractControl } from "./abstractControl";
+
+/**
+ * Type definition for a function that validates a form control and returns true if the value violates the rule.
+ *
+ * @param control {AbstractControl}   The form control to validate
+ * @returns       {boolean}           True if the value violates the rule
+ */
+export type ValidatorFn = (control: AbstractControl) => boolean;
+
+export interface ValidationRule {
+  errorMessage: string;
+  hasErrorFn: ValidatorFn;
+}
+
+/**
+ * Provides a set of validators that can be applied to an {@link AbstractControl}.
+ */
+export class Validators {
+  /**
+   * Validator that requires the form controls value to be non-empty.
+   *
+   * @param errorMessage {string}           An error message if the rule is violated
+   * @returns            {ValidationRule}   A validation rule
+   */
+  public static required(errorMessage: string): ValidationRule {
+    const hasErrorFn: ValidatorFn = (control: AbstractControl) =>
+      !control.value;
+
+    return {
+      errorMessage,
+      hasErrorFn,
+    };
+  }
+
+  /**
+   * Validator that requires the form controls value to be less than or equal to the provided number.
+   *
+   * @param errorMessage {string}           An error message if the rule is violated
+   * @param max          {number}           The max number of characters allowed
+   * @returns            {ValidationRule}   A validation rule
+   */
+  public static maxLength(errorMessage: string, max: number): ValidationRule {
+    const hasErrorFn: ValidatorFn = (control) => control.value.length > max;
+
+    return { errorMessage, hasErrorFn };
+  }
+
+  /**
+   * Validator that requires the form controls value to be strictly the length provided.
+   *
+   * @param errorMessage {string}           An error message if the rule is violated
+   * @param length       {string}           The length the value should be
+   * @returns            {ValidationRule}   A validation rule
+   */
+  public static isLength(errorMessage: string, length: number): ValidationRule {
+    const hasErrorFn: ValidatorFn = (control) =>
+      control.value.length !== length;
+
+    return { errorMessage, hasErrorFn };
+  }
+
+  /**
+   * Validator that requires the form controls value to be a valid hex id; proxies through to the {@link Validators.pattern()}.
+   *
+   * @param erroMessage {string}           An error message if the rule is violated
+   * @returns           {ValidationRule}   A validation rule
+   */
+  public static hexId(erroMessage: string): ValidationRule {
+    const hexIdRegex = /^[a-f0-9]+$/i;
+    return Validators.pattern(erroMessage, hexIdRegex);
+  }
```

Comment:
From above: 

> To reduce complexity and make the defining of the form on each page more readable, propose replacing the Validator class with just the ValidationRule[] data type, so instead of:

```typescript
beaconLocation: new FormControl(beaconLocation, [
  Validators.maxLength(
    "Where the beacon is kept has too many characters",
    250
  ),
]),
```

You’d use:

```typescript
beaconLocation: new FormControl(beaconLocation, [
{
	errorMessage:  "Where the beacon is kept has too many characters”,
	hasErrorFn: longerThan250Chars
}
```

Reply:
Good point.  It really entirely depends on what validation is required.  Most validation rules for the fields will come under either:

- Required field
- Length requirements (strict length, min length, max length)
- Some kind of Regex pattern

Really the validators class was a convenience class to reduce boilerplate code in having to specify in each file:

```typescript
{
    errorMessage: string,
    hasErrorFn: ValidatorFn,
}
```

Obviously, with that we potentially trade off readability (I've been using the validator classes from Angular for a while so appreciate the fresh pair of eyes on this!).  Like you've mentioned, I think it's worth discussing if we change the validators class to be purely functional - so just extract each of the `hasErrorFn` methods out and expose then as functions in the file and each `errorMessage` is defined in the array argument of the `AbstractControl`.  It should make the file smaller and all the tests should still pass (great way to check I've not messed up 😆 )

Relating to a point you made above, I don't believe we should be writing new functions for `longerThan100Chars` and `longerThan250Chars` as they are both a generalisation of `longerThanLength`.  We don't want to be repeating ourselves (DRY), but a nice approach might be to define a const in the page such as:

```typescript
const longerThan250Chars = longerThanLength(250)
```

And compose them that way?